### ISSUE TITLE
Unable to close the connection

### ISSUE THREAD

**Issue by aenonGit** [Tinder/Scarlet#48]
Hi, I was wandering whether it is possible to close the connection on purpose. Is it?

**Comment by raylee4204** [comment]
@aaronhe42

**Comment by zhxnlai** [comment]
@aenonGit @raylee4204 Yes it is possible. The WebSocket [echo demo](https://github.com/Tinder/Scarlet/tree/0.2.x/demo/src/main/java/com/tinder/app/websocket/echo) has a Login/Logout button that turns the connection on and off.

The trick is to define a [custom lifecycle](https://github.com/Tinder/Scarlet/blob/0.2.x/demo/src/main/java/com/tinder/app/websocket/echo/domain/LoggedInLifecycle.kt
) using the [LifecycleRegistry](https://github.com/Tinder/Scarlet/blob/0.2.x/scarlet/src/main/java/com/tinder/scarlet/lifecycle/LifecycleRegistry.kt) class and have the UI [modify it](https://github.com/Tinder/Scarlet/blob/0.2.x/demo/src/main/java/com/tinder/app/websocket/echo/view/EchoBotViewModel.kt#L39)

**Comment by jwarkentin** [comment]
This is pretty ridiculous. I've been Googling and pouring over the documentation, articles and the source looking for a way to disconnect/close the connection and couldn't find anything. This is the first thing to point me in the right direction finally but I still haven't figured out how to implement the simple ability to close a websocket. It shouldn't be this hard. At the very least, a little documentation would go a long way.

**Comment by kalanidhidev** [comment]
@zhxnlai without life-cycle is it possible to close the connection. Is there any straight forward to close the connection. I'm struct up with this. Could you help on this ?

**Comment by raylee4204** [comment]
Also, want to know if there's a way to force reconnect

**Comment by whoyawn** [comment]
A non-rx version would be also really helpful.

**Comment by rogerthat94** [comment]
Just create a LifecycleRegistry instance, like the following:
`LifecycleRegistry lifecycleRegistry = new LifecycleRegistry(0L);`

Then start the Lifecycle with the line:
`lifecycleRegistry.onNext(Lifecycle.State.Started.INSTANCE);`

Call .lifecycle on your Scarlet.Builder() and pass in lifecycleRegistry. If you have another lifecycle, you can pass in otherLifecycle.combineWith(lifecycleRegistry) instead. This whole thing would be: 
`webSocketInterface = new Scarlet.Builder().lifecycle(lifecycleRegistry).build().create(YourService.class);`

When you want to close the connection, just call:
`lifecycleRegistry.onNext(new Lifecycle.State.Stopped.WithReason(ShutdownReason.GRACEFUL));`. This will immediately make it happen. 

This doesn't use RxJava. It took me a while to figure out, but I hope it helps someone else.

**Comment by Nyanny** [comment]
@rogerthat94 Could u refactor it for 0.2 version?

**Comment by rogerthat94** [comment]
I haven't looked at 0.2. Sorry! 

According to [this](https://github.com/Tinder/Scarlet/issues/95#issuecomment-519295562) "0.1.x is current production version, and 0.2.x branch is still working in progress." 

I have just been using 0.1.9.

**Comment by ilyaklyukin** [comment]
@Nyanny  with 0.2.x with Dagger I have such an implementation:
`    @Provides
    @Singleton
    @JvmStatic
    fun provideScarletLifecycleRegistry(): LifecycleRegistry =
        LifecycleRegistry(ClientConfig.DISCONNECT_DELAY_MS)

    @Provides
    @Singleton
    @JvmStatic
    fun provideScarlet(
        lifecycleRegistry: LifecycleRegistry,
        httpClient: OkHttpClient,
        @Named("apiBaseUrl") baseUrl: String
    ): Scarlet = Scarlet(
        OkHttpWebSocket(
            httpClient,
            OkHttpWebSocket.SimpleRequestFactory(
                { Request.Builder().url(baseUrl).build() },
                { ShutdownReason.GRACEFUL }
            )),
        Scarlet.Configuration(
            lifecycle = lifecycleRegistry,
            backoffStrategy = LinearBackoffStrategy(ClientConfig.RECONNECT_INTERVAL_MS),
            messageAdapterFactories = listOf(GsonMessageAdapter.Factory()),
            streamAdapterFactories = listOf(CoroutinesStreamAdapterFactory()),
            debug = BuildConfig.DEBUG
        )
    )`

As long as I have a single Activity app, I've just injected lifeRegistry there:
`    @Inject
    lateinit var lifecycleRegistry: LifecycleRegistry`

and override methods:
`    override fun onResume() {
        super.onResume()

        setupViewModel()
        lifecycleRegistry.onNext(LifecycleState.Started)
    }

    override fun onPause() {
        lifecycleRegistry.onNext(LifecycleState.Stopped)
        viewModel.isConnected.removeObservers(this@MainActivity)

        super.onPause()
    }`

**Comment by sgc109** [comment]
@jwarkentin totally agree..

**Comment by Qkyyy** [comment]
> Just create a LifecycleRegistry instance, like the following: `LifecycleRegistry lifecycleRegistry = new LifecycleRegistry(0L);`
> 
> Then start the Lifecycle with the line: `lifecycleRegistry.onNext(Lifecycle.State.Started.INSTANCE);`
> 
> Call .lifecycle on your Scarlet.Builder() and pass in lifecycleRegistry. If you have another lifecycle, you can pass in otherLifecycle.combineWith(lifecycleRegistry) instead. This whole thing would be: `webSocketInterface = new Scarlet.Builder().lifecycle(lifecycleRegistry).build().create(YourService.class);`
> 
> When you want to close the connection, just call: `lifecycleRegistry.onNext(new Lifecycle.State.Stopped.WithReason(ShutdownReason.GRACEFUL));`. This will immediately make it happen.
> 
> This doesn't use RxJava. It took me a while to figure out, but I hope it helps someone else.


That worked for me, thanks! The `AndroidLifecycle.ofLifecycleOwnerForeground(application, lifecycleOwner)` doesn't handle for example fragment's onDestroy, onPause events. Just create the `lifecycleRegistry` in your fragment and pass it to scarlet builder, then in the onDestroy functions handle the `onNext()` methods.
